File: problems/14_Longest_Common_Prefix/longest_common_prefix.cpp
/**
 * @file longest_common_prefix.cpp
 * @brief Solution for LeetCode Problem 14: Longest Common Prefix
 *
 * This program finds the longest common prefix among an array of strings.
 * The approach uses prefix reduction by comparing each string with the current prefix.
 *
 * Compilation:
 *   g++ -std=c++17 longest_common_prefix.cpp -o longest_common_prefix
 */

#include <iostream>
#include <vector>
#include <string>

class Solution {
public:
    /**
     * @brief Finds the longest common prefix in an array of strings.
     *
     * @param strs Vector of input strings.
     * @return The longest common prefix string. Empty if none exists.
     */
    std::string longestCommonPrefix(std::vector<std::string>& strs) {
        if (strs.empty()) return "";

        std::string prefix = strs[0];

        for (size_t i = 1; i < strs.size(); ++i) {
            size_t j = 0;
            while (j < prefix.size() && j < strs[i].size() && prefix[j] == strs[i][j]) {
                ++j;
            }
            prefix = prefix.substr(0, j);

            // Early termination if prefix is empty
            if (prefix.empty()) {
                break;
            }
        }

        return prefix;
    }
};

#ifdef RUN_LOCAL_TEST
int main() {
    Solution sol;
    std::vector<std::string> test1 = {"flower", "flow", "flight"};
    std::vector<std::string> test2 = {"dog", "racecar", "car"};

    std::cout << "Test1 (Expected 'fl'): " << sol.longestCommonPrefix(test1) << std::endl;
    std::cout << "Test2 (Expected ''): " << sol.longestCommonPrefix(test2) << std::endl;

    return 0;
}
#endif

File: problems/14_Longest_Common_Prefix/longest_common_prefix.go
package main

import (
    "fmt"
)

/*
LeetCode Problem 14: Longest Common Prefix

This solution uses the **prefix reduction approach**:
- Initialize the prefix as the first string.
- Iterate through each subsequent string:
  - Reduce the prefix until it matches the current string's prefix.
  - If it becomes empty, return "" immediately.

Time Complexity: O(S), where S is the sum of all characters.
Space Complexity: O(1) extra space.
*/

// longestCommonPrefix returns the longest common prefix among an array of strings.
func longestCommonPrefix(strs []string) string {
    if len(strs) == 0 {
        return ""
    }

    prefix := strs[0]

    for _, s := range strs[1:] {
        // Reduce prefix while it does not match current string
        for len(prefix) > 0 && (len(s) < len(prefix) || s[:len(prefix)] != prefix) {
            prefix = prefix[:len(prefix)-1]
        }

        if prefix == "" {
            return ""
        }
    }

    return prefix
}

// Example usage for local testing
func main() {
    fmt.Println(longestCommonPrefix([]string{"flower", "flow", "flight"})) // Output: "fl"
    fmt.Println(longestCommonPrefix([]string{"dog", "racecar", "car"}))    // Output: ""
}
